src/renderer/contexts/GlobalNodeState.tsx

```diff
@@ -937,6 +944,63 @@ export const GlobalProvider = memo(
                     return invalid('Cannot create a connection to/from an iterator in this way.');
                 }
 
+                if (sourceNode.type === 'newIterator' && targetNode.type === 'newIterator') {
+                    return invalid('Cannot connect two iterators.');
+                }
+
+                // Connections cannot be made if:
+                // - The source has iterator lineage AND the target has iterator lineage
+                // Connects CAN be made if:
+                // - the source has iterator lineage AND the target does not
+                // - the target has iterator lineage AND the source does not
+                // - neither the source nor the target have iterator lineage
+                // Iterator lineage is defined as a node have some downstream or upstream connection to a newIterator node
+
+                const sourceDownstreamIterNodes = gatherDownstreamIteratorNodes(
+                    sourceNode,
+                    nodes,
+                    edges
+                );
+                const sourceUpstreamIterNodes = gatherUpstreamIteratorNodes(
+                    sourceNode,
+                    nodes,
+                    edges
+                );
+
+                const targetDownstreamIterNodes = gatherDownstreamIteratorNodes(
+                    targetNode,
+                    nodes,
+                    edges
+                );
+                const targetUpstreamIterNodes = gatherUpstreamIteratorNodes(
+                    targetNode,
+                    nodes,
+                    edges
+                );
+
+                const sourceHasIteratorLineage =
+                    sourceDownstreamIterNodes.size > 0 ||
+                    sourceUpstreamIterNodes.size > 0 ||
+                    sourceNode.type === 'newIterator';
+                const targetHasIteratorLineage =
+                    targetDownstreamIterNodes.size > 0 ||
+                    targetUpstreamIterNodes.size > 0 ||
+                    targetNode.type === 'newIterator';
+
+                const sourceAndTargetShareSameLineage =
+                    sourceDownstreamIterNodes.has(targetNode.id) ||
+                    sourceUpstreamIterNodes.has(targetNode.id) ||
+                    targetDownstreamIterNodes.has(sourceNode.id) ||
+                    targetUpstreamIterNodes.has(sourceNode.id);
+
+                if (
+                    sourceHasIteratorLineage &&
+                    targetHasIteratorLineage &&
+                    !sourceAndTargetShareSameLineage
+                ) {
+                    return invalid('Cannot connect two nodes with unrelated iterator lineage.');
+                }
```

Comment:
This is just wrong. Downstream nodes have nothing to do with the lineage of a node.

Right now (with our limited iterator system), it's very easy to figure out the lineage of a node. Let's represent lineage as the node id of an iterator node. Then the lineage of any node can calculate as follows:
```js
function getLineage(node): NodeId | undefined {
  if (isIterator(node)) return node.id;
  if (isCollector(node)) return undefined;
  assert(isRegularNode(node));
  return node.inputNodes.find(n => getLineage(n))
}
```

Reply:
I don't think this is correct. If you don't include downstream nodes, then you would be able to attach an iterator node to a node that is attached to an iterator node